### PR TITLE
Fixed picker reuse request body bug; Ensure all RoundTripper close request body on error cases.

### PR DESCRIPTION
**Pull request by bwplotka**
cc @ianbillett

Without this *every* HTTP requests with some kind of Body will fail if first target choosen within backends is unhealthy. 

The reason is that RoundTrip interface states clearly that no req.Body should be returned non-closed from any RoundTrip. However `lbtransport` retries same request that returns (in certain case) from `http.Transport`, so if it has `Body` and it goes to second `http.Transport.RoundTrip` invoke it returns error `error handler double close` or `http: invalid Read on closed Body`

I used TDD logic to fix it: 

I changed `integration_test.go` to reproduce the issue by:
* adding broken target to all backends
* adding non nil Body to request

This exactly reproduced issue seen on production.

The fix:
in `lbtransport` if body is non nil wraps it with `lazyBufferedReader`. It masks close so responsibility of Body close is on `lbtransport` AND every Read is buffered. To make sure the buffer is not kept if not uncecessary we do it in lazy way. Obviously reader is rewinded before each invoke of RT to ensure same view. 

Worth to note that current http.Transport implementation does not really need rewind as no body is read before dial. However we should not assume that it will never change IMO. 

Signed-off-by: Bartek Plotka <bwplotka@gmail.com>

### CODE REVIEW

File: pkg/http/tripperware/tripperware.go [hunk 1]
```diff
@@ -34,3 +35,10 @@ func DefaultWithTransport(transport *http.Transport, config *tls.Config) http.Ro
 	transport.TLSClientConfig = config
 	return &defaultTripper{Transport: transport}
 }
+
+func Close(r io.Closer) error {
```
Comment (devnev) [inline pkg/http/tripperware/tripperware.go]: An alternative would be to have a slightly more complex helper like this:
```
func toNext(req *http.Request, rt http.RoundTripper) (http.RoundTripper, func()) {
	if req.Body == nil {
		return rt, func() {}
	}
	forwarded := false
	return httpwares.RoundTripperFunc(func(actualReq *http.Request) (*http.Response, error) {
		if actualReq.Body == req.Body {
			forwarded = true
		}
		return rt.RoundTrip(actualReq)
	}), func() {
		if !forwarded {
			_ = req.Body.Close()
		}
	}
}
```

used like this in a tripperware:
```
func (t tripper) RoundTrip(req *http.Request) (*http.Response, error) {
  next, close := toNext(req)
  defer close()
  ...
  return next(req)
}
```

Comment (bwplotka) [reply pkg/http/tripperware/tripperware.go]: Agree, also I don't like this function. For something that small it should be just copied also named `closedIfNotNil` or something

Comment (bwplotka) [reply pkg/http/tripperware/tripperware.go]: your helper makes sense just maybe `ensureBodyClosed` would be more precise name?

File: pkg/kedge/http/lbtransport/body.go [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+			return n, nil
+		}
+	}
+
+	tmp := bytes.NewBuffer(b.buf)
+	n64, err := tmp.ReadFrom(io.LimitReader(b.wrapped, int64(len(p)-n)))
+	defer func() { n += int(n64) }()
+
+	if n64 == 0 {
+		return 0, io.EOF
```
Comment (Helcaraxan) [inline pkg/kedge/http/lbtransport/body.go]: Multiple points here:
- `n64 == 0` does not mean that `n == 0` hence we might still be returning non-zero-length content that has already been read into `p` at this point.
- We might not be passing through a non-EOF error here. As [I understand it](https://golang.org/src/io/io.go?s=14885:14942#L431) the `io.LimitReader` will pass through any error from the underlying reader and while `n64` might still be set to 0.

My suggestion would be to invert the condition and move the 3 lines of code below into the `if` body. This has the advantages of:
1. Allowing to simply increment `n += int(n64)` without relying on a `defer`.
2. Relying on a single `return n, err` at the end of the function.

File: pkg/kedge/http/lbtransport/body.go [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+			return n, err
+		}
+		if n == len(p) {
+			return n, nil
+		}
+	}
+
+	tmp := bytes.NewBuffer(b.buf)
+	n64, err := tmp.ReadFrom(io.LimitReader(b.wrapped, int64(len(p)-n)))
+	defer func() { n += int(n64) }()
```
Comment (Helcaraxan) [inline pkg/kedge/http/lbtransport/body.go]: Relying here on the (guaranteed) stack-ordered execution of `defer`s is somewhat itching me as it makes the code less readable. See my comment below for a potential alternative.

File: pkg/kedge/http/lbtransport/body.go [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+	return n, err
+}
+
+// newLazyBufferedReader returns lazyBufferedReader that lazely read from underlying reader if needed.
+// In worst case scenario all content from src is buffered, in the the best, nothing.
+func newLazyBufferedReader(src io.Reader) *lazyBufferedReader {
+	if src == nil {
+		return nil
+	}
+
```
Comment (Helcaraxan) [inline pkg/kedge/http/lbtransport/body.go]: Nit: spurious empty line.

File: pkg/kedge/http/lbtransport/body.go [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+}
+
+func (b *lazyBufferedReader) Read(p []byte) (n int, err error) {
+	if b == nil {
+		return 0, io.EOF
+	}
+	defer func() { b.offset += n }()
+
+	if len(b.buf)-b.offset > 0 {
+		n, err = bytes.NewReader(b.buf).Read(p)
```
[the remark this answers is not in the document]
Comment (bwplotka) [reply pkg/kedge/http/lbtransport/body.go]: Very good point, will add tests against it as well

File: pkg/http/tripperware/tripperware.go (hunk 1 above)
Comment (devnev) [reply pkg/http/tripperware/tripperware.go]: Yep naming was a bit quick'n'dirty. `EnsureBodyClosed` sounds ok.

File: pkg/kedge/http/lbtransport/body.go (hunk 2 above)
Comment (bwplotka) [reply pkg/kedge/http/lbtransport/body.go]: Good catch!

File: pkg/kedge/http/lbtransport/body.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (Helcaraxan) [reply pkg/kedge/http/lbtransport/body.go]: This was not caught in your testing as the pre-condition for it to occur is a read error on the wrapped reader and some weird use-pattern.

File: pkg/kedge/http/lbtransport/transport.go [hunk 6]
```diff
@@ -111,15 +113,25 @@ func (s *tripper) RoundTrip(r *http.Request) (*http.Response, error) {
...
 	if len(targetsRef) == 0 {
 		err := errors.Errorf("lb: no backend is available for %s. 0 resolved addresses.", s.targetName)
 		reporter.Extract(r).ReportError(errtypes.NoResolutionAvailable, err)
+		_ = tripperware.Close(r.Body)
 		return nil, err
 	}
 
+	if r.Body != nil {
+		// We have to own true body for the request because we cannot reuse same reader closer
+		// in multiple calls to http.Transport.
```
Comment (devnev) [inline pkg/kedge/http/lbtransport/transport.go]: Isn't this what `http.Request`'s `GetBody` is for? i.e. use `GetBody` if available, replace with lazy buffer thingy if GetBody unavailable? might be too complex for now.

File: pkg/kedge/http/lbtransport/body.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (Helcaraxan) [reply pkg/kedge/http/lbtransport/body.go]: Actually strip that: you can trigger it by rewinding and then reading less than the combined length of all previously read content.

[the remark this answers is not in the document]
Comment (bwplotka) [reply pkg/kedge/http/lbtransport/body.go]: yea

File: pkg/kedge/http/lbtransport/transport.go (hunk 6 above)
Comment (bwplotka) [reply pkg/kedge/http/lbtransport/transport.go]: This makes sense I think

Comment (bwplotka) [reply pkg/kedge/http/lbtransport/transport.go]: But we still have to have reference to `*lazyBufferedReader` to rewind so let's maybe stick to this now.

File: pkg/kedge/http/lbtransport/body_test.go [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+import (
+	"bytes"
+	"io"
+	"testing"
+
+	"github.com/stretchr/testify/require"
+)
+
+func TestLazyBufferedReader(t *testing.T) {
+	for _, tcase := range []struct {
```
Comment (Helcaraxan) [inline pkg/kedge/http/lbtransport/body_test.go]: Given the number of test cases I'd suggest adding names to each of them that can be used to run the corresponding sub-test. It also allows people to easily check whether a given edge-cases they've come up with is actually tested for.

File: pkg/kedge/http/lbtransport/transport.go [hunk 8]
```diff
@@ -111,15 +113,25 @@ func (s *tripper) RoundTrip(r *http.Request) (*http.Response, error) {
...
 
 	if len(targetsRef) == 0 {
 		err := errors.Errorf("lb: no backend is available for %s. 0 resolved addresses.", s.targetName)
 		reporter.Extract(r).ReportError(errtypes.NoResolutionAvailable, err)
+		_ = tripperware.Close(r.Body)
 		return nil, err
 	}
 
+	if r.Body != nil {
+		// We have to own true body for the request because we cannot reuse same reader closer
```
Comment (Helcaraxan) [inline pkg/kedge/http/lbtransport/transport.go]: Nit: `s/true/the`?

File: pkg/kedge/http/lbtransport/body.go [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+		b.offset += n
+	}
+
+	var n64 int64
+	if err == nil && n < len(p) {
+		// Try to buffer rest (if needed) from wrapped io.Reader.
+		tmp := bytes.NewBuffer(b.buf)
+		n64, err = tmp.ReadFrom(io.LimitReader(b.wrapped, int64(len(p)-n)))
+		if n64 > 0 {
+			b.buf = tmp.Bytes()
```
[the remark this answers is not in the document]
Comment (bwplotka) [reply pkg/kedge/http/lbtransport/body.go]: I agree, I am hacking bit here

File: pkg/http/tripperware/tripperware.go (hunk 1 above)
Comment (bwplotka) [reply pkg/http/tripperware/tripperware.go]: So we have multiple tripperwares. Like 6-7. With this we essentially double the stack frame as single tripperware is now actually 2 - the tripperware + EnsureBodyClosed. Even though we don't care about number of tripperwares this feels wrong. Maybe some wrapper over body in single place to close it if not closed on exit path?

Comment (devnev) [reply pkg/http/tripperware/tripperware.go]: Wrapping body cleanly is hard, as the concrete type may implement various interfaces for optimisation (readerfrom, writerto). I don't get the same sense of this feeling wrong, can you elaborate why?

File: pkg/kedge/http/lbtransport/body.go [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+			copy(p[n:], b.buf[b.offset:])
+			n += int(n64)
+			b.offset += int(n64)
+		}
+	}
+
+	// Buffer.ReadFrom masks io.EOF so we assume EOF once n == 0 and no error.
```
[the remark this answers is not in the document]
Comment (bwplotka) [reply pkg/kedge/http/lbtransport/body.go]: Good point

File: pkg/http/tripperware/tripperware.go (hunk 1 above)
Comment (bwplotka) [reply pkg/http/tripperware/tripperware.go]: I decided to just go for simple explicit closes, what do you think? I just think it's too much code for something that simple ):

File: pkg/kedge/http/lbtransport/body.go [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+	if err == nil && n == 0 && len(p) > 0 {
+		return 0, io.EOF
+	}
+	return n, err
+}
+
+// newLazyBufferedReader returns replayableReader that lazely read from underlying reader if needed.
+// In worst case scenario all content from src is buffered, in the the best, nothing.
```
Comment (Helcaraxan) [inline pkg/kedge/http/lbtransport/body.go]: This last sentence of the comment is not exactly true. Suggestion:
```go
// The content read from the source is buffered in a lazy fashion to keep storage requirements
// limited to a minimum while still allowing for the reader to be rewinded and previously read
// content to be replayed.
```

File: pkg/kedge/http/lbtransport/body.go [hunk 12]
```diff
@@ -0,0 +1,64 @@
...
+func newLazyBufferedReader(src io.Reader) *replayableReader {
```
Comment (Helcaraxan) [inline pkg/kedge/http/lbtransport/body.go]: Nit: This also needs to be renamed after we renamed `lazyBufferedReader` to `replayableReader`.

File: pkg/kedge/http/lbtransport/body_test.go [hunk 13]
```diff
@@ -0,0 +1,113 @@
...
+		name string
+		src                 io.Reader
+		sequentialReadBytes []int
+		rewindBeforeRead    []bool
+
+		expectedBytes [][]byte
+		expectedErrs  []error
+	}{
+		{
+			name: "wrapped nil, read should return EOF",
```
Comment (Helcaraxan) [inline pkg/kedge/http/lbtransport/body_test.go]: Nit: Given that these names are used as test names we might want to make them adhere to the usual style with a camel-cased sentence as a single word?

File: pkg/http/tripperware/tripperware.go [hunk 14]
```diff
@@ -34,3 +35,9 @@ func DefaultWithTransport(transport *http.Transport, config *tls.Config) http.Ro
 	transport.TLSClientConfig = config
 	return &defaultTripper{Transport: transport}
 }
+
+func closeIfNotNil(r io.Closer) {
```
Comment (Helcaraxan) [inline pkg/http/tripperware/tripperware.go]: Is there a specific reason why we are calling this `closeIfNotNil` instead of the `ensureClosed` that is used in the identical function in the `lbtransport`package?

I personally have a strong preference for `ensureClosed`.

Comment (bwplotka) [reply pkg/http/tripperware/tripperware.go]: Mistake. Will get all to `ensureClosed`

Comment (devnev) [reply pkg/http/tripperware/tripperware.go]: I don't think `ensureClosed` describes what this function is doing - I prefer `closeIfNotNil`, and even better would be a `closeBody(r *http.Request)` function.

File: pkg/http/tripperware/tripperware.go (hunk 1 above)
Comment (devnev) [reply pkg/http/tripperware/tripperware.go]: To me, the explicit closes feel significantly more error-prone than the other helper, so more likely to be gotten wrong in future changes. Consider that we got this wrong at first, and that future developers will need to read through the http request docs the same way we have to understand the rationale for the close calls. Maybe there's another way, maybe `closeBody(*http.Request)` will make it clearer?

File: pkg/http/tripperware/tripperware.go (hunk 14 above)
Comment (bwplotka) [reply pkg/http/tripperware/tripperware.go]: Talked to @Helcaraxan offline - moving to `closeIfNotNil`
